form/data_mappers.rst

```diff
@@ -0,0 +1,197 @@
+data transformer. Mapping this ``DateTime`` instance as value of a property on
+the entity is done by a data mapper.
+
+Creating a Data Mapper
+----------------------
+
+Assume you're saving a set of colors in the database. For this, you're using an
+immutable color object::
+
+    // src/AppBundle/Colors/Color.php
+    namespace AppBundle\Colors;
+
+    class Color
+    {
+        private $red;
+        private $green;
+        private $blue;
+
+        public function __construct($red, $green, $blue)
+        {
+            $this->red = $red;
+            $this->green = $green;
+            $this->blue = $blue;
+        }
+
+        public function getRed()
+        {
+            return $this->red;
+        }
+
+        public function getGreen()
+        {
+            return $this->green;
+        }
+
+        public function getBlue()
+        {
+            return $this->blue;
+        }
+    }
+
+The form type should be allowed to edit a color. As the color object is
+immutable, a new color object has to be created each time one of the values is
+changed.
+
+.. tip::
+
+    If you're using a mutable object with constructor arguments, instead of
+    using a data mapper, you should configure the ``empty_data`` with a closure
+    as described in
+    :ref:`How to Configure empty Data for a Form Class <forms-empty-data-closure>`.
+
+The red, green and blue form fields have to be mapped to the constructor
+arguments and the ``Color`` instance has to be mapped to red, green and blue
+form fields. Recognize a familair pattern? It's time for a data mapper!
+
+.. code-block:: php
+
+    // src/AppBundle/Form/DataMapper/ColorMapper.php
+    namespace AppBundle\Form\DataMapper;
+
+    use AppBundle\Colors\Color;
+    use Symfony\Component\Form\DataMapperInterface;
+    use Symfony\Component\Form\Exception\UnexpectedTypeException;
+
+    class ColorMapper extends DataMapperInterface
+    {
+        public function mapDataToForms($data, $forms)
+        {
+            // there is no data yet, a new color will be created
+            if (null === $data) {
+                return;
+            }
+
+            // invalid data type, this message will not be shown to the user (see below)
+            if (!$data instanceof Color) {
+                throw new UnexpectedTypeException($data, Color::class);
+            }
+
+            $forms = iterator_to_array($forms);
+
+            // set form field values
+            $forms['red']->setData($data->getRed());
+            $forms['green']->setData($data->getGreen());
+            $forms['blue']->setData($data->getBlue());
+        }
+
+        public function mapFormsToData($forms, &$data)
+        {
+            $forms = iterator_to_array($forms);
+
+            // get form field values
+            $red = $forms['red']->getData();
+            $green = $forms['green']->getData();
+            $blue = $forms['blue']->getData();
+
+            $data = new Color($red, $green, $blue);
```

Comment:
woh! I pass by reference so rarely, that I'm honestly shocked that you can simply override the `$data` variable and this works! If so (I'm sure it does - Bernhard's blog post uses this same method), maybe we should add a quick comment:

> // Overriding $data is enough, since it is passed by reference.